### ISSUE TITLE
inference issues with german fine-tune

### ISSUE THREAD

**Issue by christophschuhmann** [canopyai/Orpheus-TTS#114]
Here is a very nice German fine tune that works well with short sentences:

https://huggingface.co/SebastianBodza/Kartoffel_Orpheus-3B_german_natural-v0.1

Unfortunately, the issue doesn't stop with my script. After generating the speech it was prompted to say, it starts mumbling and hallucinating. I told  O1 to stop when either the end-of-text or end-of-audio token was hit, but I'm not an inference expert, so I don't know why it keeps generating. By the way, at the moment, the script splits long texts into short segments and later concatenates them into one file. So, when you run this, you'll hear a proper sentence, then mumbling, and then another proper sentence and more mumbling, and so on, because it's actually pretty good at the beginning of each generation.

# -*- coding: utf-8 -*-
"""
tts_pipeline.py

Dieses Skript lädt einmalig das Modell "SebastianBodza/Kartoffel_Orpheus-3B_german_natural-v0.1"
sowie den SNAC-Audio-Decoder ("hubertsiuzdak/snac_24khz") und bietet eine Funktion 
`generate_tts_file(...)` für die TTS-Generierung.

Wichtige Punkte:
  - Die Generierung stoppt, sobald **entweder** ein EndOfText-Token oder ein EndOfAudio-Token 
    erkannt wird (welches früher auftaucht).
  - Der Text kann entweder als langer String oder als Liste von kurzen Sätzen übergeben werden.
  - Bei langen Texten (>= 400 Zeichen) wird dieser in Sätze (mittels nltk) und 
    anschließend in Wortgruppen (je 10–30 Wörter) unterteilt.
  - Zwischen einzelnen Audio-Segmenten wird eine Stille von 300 ms eingefügt.
  
Benötigte Pakete:
  pip install torch transformers pydub snac huggingface_hub nltk

ffmpeg muss systemweit installiert sein.
"""

import os
import re
import torch
import numpy as np

from transformers import AutoConfig, AutoModelForCausalLM, AutoTokenizer
from snac import SNAC
from pydub import AudioSegment

# nltk-Tokenizer
import nltk
try:
    nltk.data.find('tokenizers/punkt')
except LookupError:
    nltk.download('punkt')

# ------------------------------------------------------------------------------
#                       GLOBALE KONFIGURATION & PARAMETER
# ------------------------------------------------------------------------------
HF_TOKEN = "xxxxxxxxx"  # Bitte ggf. anpassen
MODEL_ID = "SebastianBodza/Kartoffel_Orpheus-3B_german_natural-v0.1"

# Audio-Sample-Rate
SAMPLE_RATE = 24000

# Standardwerte für das Text-Splitting
CHAR_THRESHOLD = 400   # Ab dieser Zeichenlänge wird gesplittet
MIN_WORDS = 10         # Minimal 10 Wörter pro Segment
MAX_WORDS = 30         # Maximal 30 Wörter pro Segment

# Standard-Inferenz-Parameter
DEFAULT_TEMPERATURE        = 0.6
DEFAULT_TOP_P              = 0.95
DEFAULT_REPETITION_PENALTY = 1.1
MAX_NEW_TOKENS             = 800

# Stille zwischen Segmenten (Millisekunden)
SEGMENT_SILENCE_MS = 300

# Default-Output-Ordner
DEFAULT_OUTPUT_DIR = "tts_outputs"

# ------------------------------------------------------------------------------
#                  EINMALIGES LADEN DES MODELLS / TOKENIZERS / DECODERS
# ------------------------------------------------------------------------------
print(f"Lade Modell-Konfiguration von: {MODEL_ID}")
config = AutoConfig.from_pretrained(MODEL_ID, token=HF_TOKEN)

# rope_scaling-Fix (von "rope_type" zu "type")
if hasattr(config, "rope_scaling") and isinstance(config.rope_scaling, dict):
    if "rope_type" in config.rope_scaling:
        config.rope_scaling["type"] = config.rope_scaling.pop("rope_type")

print("Lade Modell (einmalig) ...")
model = AutoModelForCausalLM.from_pretrained(
    MODEL_ID,
    config=config,
    token=HF_TOKEN,
    torch_dtype=torch.bfloat16,
)
model.cuda()
model.eval()
print("Modell erfolgreich geladen.")

print("Lade Tokenizer ...")
tokenizer = AutoTokenizer.from_pretrained(MODEL_ID, token=HF_TOKEN)
print("Tokenizer geladen.")

print("Lade SNAC-Audio-Decoder (einmalig) ...")
snac_model = SNAC.from_pretrained("hubertsiuzdak/snac_24khz")
snac_device = torch.device("cpu")
snac_model.to(snac_device)
print("SNAC auf CPU geladen.")

# ------------------------------------------------------------------------------
#                       SPEZIELLE TOKEN-DEFINITIONEN
# ------------------------------------------------------------------------------
TOKEN_START_HEADER = 128259  # SOH (Start-of-Header)
TOKEN_END_OF_TEXT  = 128009  # EOT (End-of-Text)
TOKEN_END_OF_HDR   = 128260  # EOH (End-of-Header)
TOKEN_START_AUDIO  = 128257  # Start-of-Audio
TOKEN_END_AUDIO    = 128258  # End-of-Audio
TOKEN_AUDIO_OFFSET = 128266  # Audio-Code Offset

# ------------------------------------------------------------------------------
#        OPTIONALES CLEANING FÜR DEUTSCHE TEXTE (Umlaute, e.V., ß erlaubt)
# ------------------------------------------------------------------------------
def clean_text_for_german(text: str) -> str:
    """
    Erlaubt Umlaute (Ä, Ö, Ü, ä, ö, ü), ß, Ziffern, Satzzeichen und Leerzeichen.
    "e.V." wird korrekt mit Punkten geschrieben. Exotische Sonderzeichen werden entfernt.
    """
    # Beispiel: "Lion eV" zu "Lion e.V."
    text = text.replace("Lion eV", "Lion e.V.")
    text = text.replace("Lion EV", "Lion e.V.")
    text = text.replace("Lion e.v", "Lion e.V.")
    
    pattern = r"[^a-zA-Z0-9äöüÄÖÜß.,!?:;\-–\(\)\"' ]"
    cleaned = re.sub(pattern, "", text)
    cleaned = re.sub(r"\s+", " ", cleaned).strip()
    return cleaned

# ------------------------------------------------------------------------------
#                       TEXTSPLITTING (nltk)
# ------------------------------------------------------------------------------
def split_text_into_sentences_de(text: str):
    """
    Splittet den Text mit nltk in deutsche Sätze.
    """
    return nltk.sent_tokenize(text, language='german')

def chunk_sentence_into_wordgroups(sentence: str, min_words: int, max_words: int):
    """
    Teilt einen Satz in Wortgruppen von min_words bis max_words.
    """
    words = sentence.split()
    chunks = []
    i = 0
    while i < len(words):
        end = min(i + max_words, len(words))
        if (end - i) < min_words and end < len(words):
            end = i + min_words
        chunk = words[i:end]
        chunks.append(" ".join(chunk))
        i = end
    return chunks

def split_if_long(
    text: str,
    char_threshold: int = CHAR_THRESHOLD,
    min_words: int = MIN_WORDS,
    max_words: int = MAX_WORDS
):
    """
    Wenn text >= char_threshold Zeichen hat:
      1) Zuerst in Sätze splitten.
      2) Dann jeden Satz in Wortgruppen (min_words bis max_words).
    Andernfalls: Rückgabe als [text].
    """
    text = text.strip()
    if len(text) < char_threshold:
        return [text]

    sents = split_text_into_sentences_de(text)
    final_chunks = []
    for s in sents:
        s = s.strip()
        if not s:
            continue
        partials = chunk_sentence_into_wordgroups(s, min_words, max_words)
        final_chunks.extend(partials)

    return final_chunks

# ------------------------------------------------------------------------------
#      HILFSFUNKTION: AUDIO-CODES AUF 3 SNAC-SCHICHTEN VERTEILEN
# ------------------------------------------------------------------------------
def redistribute_codes(codes_):
    """
    Konvertiert Gruppen von 7 Tokens in das 3-Schichten-Format von SNAC.
    """
    if len(codes_) == 0 or (len(codes_) % 7) != 0:
        return None
    l1, l2, l3 = [], [], []
    groups = len(codes_) // 7
    for g in range(groups):
        base = 7 * g
        try:
            l1.append(codes_[base + 0])
            l2.append(codes_[base + 1] - 4096)
            l3.append(codes_[base + 2] - 2 * 4096)
            l3.append(codes_[base + 3] - 3 * 4096)
            l2.append(codes_[base + 4] - 4 * 4096)
            l3.append(codes_[base + 5] - 5 * 4096)
            l3.append(codes_[base + 6] - 6 * 4096)
        except IndexError:
            return None
    t1 = torch.tensor(l1, dtype=torch.long).unsqueeze(0).to(snac_device)
    t2 = torch.tensor(l2, dtype=torch.long).unsqueeze(0).to(snac_device)
    t3 = torch.tensor(l3, dtype=torch.long).unsqueeze(0).to(snac_device)
    return [t1, t2, t3]

# ------------------------------------------------------------------------------
#   GENERIERT EIN AUDIOSEGMENT (pydub) FÜR EINEN TEXTPROMPT (SPEAKER: TEXT)
# ------------------------------------------------------------------------------
def generate_audio_for_text(
    text_prompt: str,
    temperature: float = DEFAULT_TEMPERATURE,
    top_p: float = DEFAULT_TOP_P,
    repetition_penalty: float = DEFAULT_REPETITION_PENALTY,
    max_new_tokens: int = MAX_NEW_TOKENS
) -> AudioSegment:
    """
    Führt die Audio-Generierung für einen Prompt durch.
    Die Generierung stoppt, sobald **entweder** TOKEN_END_OF_TEXT oder TOKEN_END_AUDIO erscheint.
    
    Rückgabe: pydub.AudioSegment oder None, falls nichts generiert wurde.
    """
    t = tokenizer(text_prompt, return_tensors="pt").input_ids
    new_ids = torch.cat([
        torch.tensor([[TOKEN_START_HEADER]]),
        t,
        torch.tensor([[TOKEN_END_OF_HDR]])
    ], dim=1).cuda()

    with torch.no_grad():
        generated = model.generate(
            input_ids=new_ids,
            max_new_tokens=max_new_tokens,
            do_sample=True,
            temperature=temperature,
            top_p=top_p,
            repetition_penalty=repetition_penalty,
            # Stoppt sobald einer der beiden Tokens auftaucht:
            eos_token_id=[TOKEN_END_OF_TEXT, TOKEN_END_AUDIO],
            pad_token_id=(tokenizer.pad_token_id or TOKEN_END_AUDIO)
        )

    generated_cpu = generated[0].cpu()
    starts = (generated_cpu == TOKEN_START_AUDIO).nonzero(as_tuple=True)[0]
    if len(starts) == 0:
        return None

    last_start_idx = starts[-1].item()
    after_start = generated_cpu[last_start_idx + 1:]
    ends = (after_start == TOKEN_END_AUDIO).nonzero(as_tuple=True)[0]
    if len(ends) > 0:
        first_end_idx = ends[0].item()
        audio_tokens = after_start[:first_end_idx]
    else:
        audio_tokens = after_start

    adjusted = [x.item() - TOKEN_AUDIO_OFFSET for x in audio_tokens if x.item() >= TOKEN_AUDIO_OFFSET]
    length_7 = (len(adjusted) // 7) * 7
    codes = adjusted[:length_7]
    code_tensors = redistribute_codes(codes)
    if code_tensors is None:
        return None

    with torch.no_grad():
        audio_hat = snac_model.decode(code_tensors)

    audio_np = audio_hat.squeeze().cpu().numpy()
    if len(audio_np) == 0:
        return None

    max_val = np.max(np.abs(audio_np))
    if max_val > 1e-9:
        audio_np /= max_val

    audio_int16 = np.int16(audio_np * 32767)
    seg = AudioSegment(
        audio_int16.tobytes(),
        frame_rate=SAMPLE_RATE,
        sample_width=2,
        channels=1
    )
    return seg

# ------------------------------------------------------------------------------
#   HAUPTFUNKTION: TTS FÜR LÄNGERE TEXTE / LISTEN VON SÄTZEN UND EXPORT ALS MP3
# ------------------------------------------------------------------------------
def generate_tts_file(
    text_or_sentences,
    speaker: str,
    out_path: str,
    temperature: float = DEFAULT_TEMPERATURE,
    top_p: float = DEFAULT_TOP_P,
    repetition_penalty: float = DEFAULT_REPETITION_PENALTY,
    max_new_tokens: int = MAX_NEW_TOKENS,
    char_threshold: int = CHAR_THRESHOLD,
    min_words: int = MIN_WORDS,
    max_words: int = MAX_WORDS,
    segment_silence_ms: int = SEGMENT_SILENCE_MS
) -> str:
    """
    Nimmt entweder:
      - einen langen String (wird intern in Sätze/Wortgruppen aufgeteilt, sofern >= char_threshold)
      - eine Liste bereits kurzer Sätze.
    Reinigt den Text (beibehaltend deutsche Sonderzeichen wie Umlaute, ß und "e.V."), 
    generiert für jeden Segment den Prompt "SPEAKER: <Segment>" und hängt die Audios mit
    segment_silence_ms Stille zusammen.
    
    Speichert das Ergebnis als MP3 an out_path und gibt diesen Pfad zurück.
    """
    segments = []
    if isinstance(text_or_sentences, str):
        cleaned_text = clean_text_for_german(text_or_sentences)
        splitted = split_if_long(cleaned_text, char_threshold=char_threshold, min_words=min_words, max_words=max_words)
        segments = splitted
    elif isinstance(text_or_sentences, list):
        for s in text_or_sentences:
            c = clean_text_for_german(s)
            parts = split_if_long(c, char_threshold=char_threshold, min_words=min_words, max_words=max_words)
            segments.extend(parts)
    else:
        raise ValueError("text_or_sentences muss ein String oder eine Liste von Strings sein.")

    combined_segment = AudioSegment.empty()
    for idx, chunk_text in enumerate(segments, start=1):
        prompt = f"{speaker}: {chunk_text}"
        seg_chunk = generate_audio_for_text(
            prompt,
            temperature=temperature,
            top_p=top_p,
            repetition_penalty=repetition_penalty,
            max_new_tokens=max_new_tokens
        )
        if seg_chunk is None:
            continue

        if len(combined_segment) > 0:
            combined_segment += AudioSegment.silent(duration=segment_silence_ms)
        combined_segment += seg_chunk

    if len(combined_segment) > 0:
        os.makedirs(os.path.dirname(out_path) or ".", exist_ok=True)
        combined_segment.export(out_path, format="mp3", bitrate="128k")
        print(f"Gespeichert: {out_path}")
    else:
        print("Keine Audiodaten generiert. Überspringe Export.")

    return out_path

# ------------------------------------------------------------------------------
#            BEISPIEL-VERWENDUNG: LISTE KURZER SÄTZE
# ------------------------------------------------------------------------------
if __name__ == "__main__":
    example_sentences = [
        "Hallo, ich bin SchoolBuddy.",
        "Ich bin ein Sprachassistent von Lion e.V.",
        "Heute möchte ich dir etwas erzählen über die Zukunft des KI-gestützten Lernens.",
        "Stell dir vor, es fühlt sich an wie ein echtes Telefongespräch.",
        "Du nimmst dein Gerät in die Hand und sprichst einfach drauflos."
    ]

    speaker = "Sophie"
    out_file = os.path.join(DEFAULT_OUTPUT_DIR, "demo_sophie_list_sentences.mp3")

    generate_tts_file(
        text_or_sentences=example_sentences,
        speaker=speaker,
        out_path=out_file,
        temperature=0.6,
        top_p=0.9,
        repetition_penalty=1.1
    )

    print("Fertig.")

**Comment by stlohrey** [comment]
I tested the model with single sentence inference and it works just fine, did you try this on your end?

**Comment by christophschuhmann** [comment]
Can you share your code? :)

**Comment by stlohrey** [comment]
i just used the inference part of the unsloth colab, https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Orpheus_(3B)-TTS.ipynb
